3662-Reify-type-of-test-results

juliendelplanque opened the pull request:
Experiment to fix #3662 

Basically, I created `TestResultType` class which model the result of a test and added subclasses.
This change will make analysis made over tests easier to implement as the type of test result (pass, fail, error, etc) will be directly accessible from TestCase instances.

Additionally, it will allow one to extend the set of types available in the system easily.
Before "types" of test case execution were encoded by the collection in which they were put in a `TestResult` instance.

If this contribution is accepted, the next step is to adapt users of SUnit to rely on result type (SUnitRunner, Commandline runner, Calypso, etc). It will simply a lot the code to decide for the color of a TestResult for example as this feature will be built-in.

juliendelplanque committed:
Reified test results.

jecisc commented on src/SUnit-Core/ClassFactoryForTestCase.class.st:
```diff
@@ -271,7 +271,7 @@ ClassFactoryForTestCase >> newTraitNamed: aTraitName uses: aTraitComposition tag
 
 { #category : #creating }
-ClassFactoryForTestCase >> newTraitUsing: aTraitComposition. [
+ClassFactoryForTestCase >> newTraitUsing: aTraitComposition [.
```
```suggestion
ClassFactoryForTestCase >> newTraitUsing: aTraitComposition [
```

juliendelplanque replied:
What is that? I did not touch this.

jecisc replied:
It is probably a fix in Iceberg. When you commit it export the whole package.

juliendelplanque replied:
Ok I commit that.

jecisc commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -86,15 +86,6 @@ TestAsserter >> assert: actualNumber closeTo: expectedNumber precision: epsilon
 ]
 
-{ #category : #asserting }
-TestAsserter >> assert: actualNumber closeTo: expectedNumber precision: margin [
```
Are those removals wanted?

juliendelplanque replied:
Hum, no I do not know what happened I will check.

ErikOnBike replied:
FYI: I had exactly the same issue (same changes I did not make myself) when I updated something in the SUnit-Core package (see #3583).

ErikOnBike replied:
BTW: My solution was to manually restore things through git directly on the filesystem instead of the through the Smalltalk image.

juliendelplanque replied:
Thank you! I will revert these changes.

juliendelplanque replied:
I tried to remove the methods from Pharo (they were still there) and to re-introduce them but it does not work... What the hell...

ErikOnBike replied:
If my bash history is correct as well as my brain ;-) I did the following:
git checkout f99498e -- src/SUnit-Core/ClassFactoryForTestCase.class.st
git checkout a57a347 -- src/SUnit-Core/TestAsserter.class.st
git add src/SUnit-Core/ClassFactoryForTestCase.class.st
git add src/SUnit-Core/TestAsserter.class.st
git commit -m "Revert 2 accidentally updated classes/files"

But check on the individual steps please.

jecisc commented on src/SUnit-Core/TestFailed.class.st:
```diff
@@ -0,0 +1,18 @@
+}
+
+{ #category : #accessing }
+TestFailed >> color [
```
It could be an extension of SUnit-UI?

juliendelplanque replied:
Yes I think as well.

juliendelplanque committed:
Update src/SUnit-Core/ClassFactoryForTestCase.class.st

Co-Authored-By: CyrilFerlicot <cyril@ferlicot.me>

juliendelplanque committed:
Really removed methods removed by Iceberg apprently?
(Will reintroduce them in next commit).

juliendelplanque committed:
Re-introduced methods that magically disappeared from the git...

Ducasse commented:
Julien can you also talk to damien because we were discussing that we need something that accumulate results (and not just a composite).

Ducasse commented:
44 tests broke so I relaunch the build to verify.

jecisc committed:
Revert methods deletions.

jecisc commented:
For the removed methods here is a fix: https://github.com/juliendelplanque/pharo/pull/1

ErikOnBike commented:
It might be me, but if I look at the file for TestAsserter in Github I see two methods with the same selector. This should not be possible, should it?!

File: https://github.com/pharo-project/pharo/blob/Pharo8.0/src/SUnit-Core/TestAsserter.class.st
Line (81): TestAsserter >> assert: actualNumber closeTo: expectedNumber precision: epsilon [
Line (90): TestAsserter >> assert: actualNumber closeTo: expectedNumber precision: margin [

Might this be the cause of having these in our PRs?

jecisc commented:
Indeed. In that case, it's acceptable that this PR remove one version of each.

Maybe two people added those methods and the merge of git kept both versions?

juliendelplanque commented:
@Ducasse yes I will discuss with him.

Here I did this small step because it slows us down at improving the test infrastructure...

juliendelplanque commented:
@ErikOnBike @jecisc Maybe this is a bug in Tonel. Tonel should do check that the file has no duplicated methods inside... I guess where there are, the behaviour is unspecified.

juliendelplanque committed:
Merge pull request #1 from jecisc/3662-Reify-type-of-test-results

Revert methods deletions.

juliendelplanque commented:
I found a problem in the implementation, I will try to fix it and add unit test.

juliendelplanque committed:
Fixed bug that was errasing the #resultType.

juliendelplanque committed:
Merge fa88fdc9576812cadf2d9d1fc1b70f661f05c65a

juliendelplanque committed:
Simulate hold behaviour for failures.
They were stored in a Set to avoid getting duplicates.
Now, we use a guard clause.
This will probably need to be fixed more properly later.

juliendelplanque committed:
Implemented #hash and #=.

juliendelplanque commented:
Errors seems not related but I am not sure.

juliendelplanque committed:
Renamed TestResultType to TestResultSeverity.

juliendelplanque committed:
Moved #color methods to SUnit-UI package since they belong here.

juliendelplanque committed:
RenamedUnknownTestResult to TestUnknownResult.

juliendelplanque commented:
It should be ready now.

juliendelplanque committed:
Fixing mess with methods that were duplicated...

juliendelplanque commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -319,40 +310,22 @@ TestAsserter >> deny: aBooleanOrBlock [
-	 the default precision (see class-side method #defaultPrecisionsForCloseToComparison).
-	"
-	self deny: actualNumber closeTo: expectedNumber precision: self class defaultPrecisionsForCloseToComparison
+{ #category : #asserting }		
```
```suggestion
{ #category : #asserting }
```

juliendelplanque commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -319,40 +310,22 @@ TestAsserter >> deny: aBooleanOrBlock [
-	"
-	self deny: actualNumber closeTo: expectedNumber precision: self class defaultPrecisionsForCloseToComparison
+{ #category : #asserting }		
+TestAsserter >> deny: actualNumber closeTo: expectedNumber [		
```
```suggestion
TestAsserter >> deny: actualNumber closeTo: expectedNumber [
```

juliendelplanque commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -319,40 +310,22 @@ TestAsserter >> deny: aBooleanOrBlock [
-	self deny: actualNumber closeTo: expectedNumber precision: self class defaultPrecisionsForCloseToComparison
+{ #category : #asserting }		
+TestAsserter >> deny: actualNumber closeTo: expectedNumber [		
+	"Tell whether the actualNumber and expectedNumber are NOT close to each others with a margin of		
```
```suggestion
	"Tell whether the actualNumber and expectedNumber are NOT close to each others with a margin of
```

juliendelplanque commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -319,40 +310,22 @@ TestAsserter >> deny: aBooleanOrBlock [
+{ #category : #asserting }		
+TestAsserter >> deny: actualNumber closeTo: expectedNumber [		
+	"Tell whether the actualNumber and expectedNumber are NOT close to each others with a margin of		
+	 the default precision (see class-side method #defaultPrecisionsForCloseToComparison).		
```
```suggestion
	 the default precision (see class-side method #defaultPrecisionsForCloseToComparison).
```

juliendelplanque commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -319,40 +310,22 @@ TestAsserter >> deny: aBooleanOrBlock [
+TestAsserter >> deny: actualNumber closeTo: expectedNumber [		
+	"Tell whether the actualNumber and expectedNumber are NOT close to each others with a margin of		
+	 the default precision (see class-side method #defaultPrecisionsForCloseToComparison).		
+	"		
```
```suggestion
	"
```

juliendelplanque commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -319,40 +310,22 @@ TestAsserter >> deny: aBooleanOrBlock [
+	"Tell whether the actualNumber and expectedNumber are NOT close to each others with a margin of		
+	 the default precision (see class-side method #defaultPrecisionsForCloseToComparison).		
+	"		
+	self deny: actualNumber closeTo: expectedNumber precision: self class defaultPrecisionsForCloseToComparison		
```
```suggestion
	self deny: actualNumber closeTo: expectedNumber precision: self class defaultPrecisionsForCloseToComparison
```

juliendelplanque commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -319,40 +310,22 @@ TestAsserter >> deny: aBooleanOrBlock [
-{ #category : #asserting }
-TestAsserter >> deny: actualNumber closeTo: expectedNumber precision: epsilon [
-	"Tell whether the actualNumber and expectedNumber are NOT close to each others with a margin of epsilon."
+{ #category : #asserting }		
```
```suggestion
{ #category : #asserting }
```

juliendelplanque commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -319,40 +310,22 @@ TestAsserter >> deny: aBooleanOrBlock [
-TestAsserter >> deny: actualNumber closeTo: expectedNumber precision: epsilon [
-	"Tell whether the actualNumber and expectedNumber are NOT close to each others with a margin of epsilon."
+{ #category : #asserting }		
+TestAsserter >> deny: actualNumber closeTo: expectedNumber precision: epsilon [		
```
```suggestion
TestAsserter >> deny: actualNumber closeTo: expectedNumber precision: epsilon [
```

juliendelplanque commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -319,40 +310,22 @@ TestAsserter >> deny: aBooleanOrBlock [
-	"Tell whether the actualNumber and expectedNumber are NOT close to each others with a margin of epsilon."
+{ #category : #asserting }		
+TestAsserter >> deny: actualNumber closeTo: expectedNumber precision: epsilon [		
+	"Tell whether the actualNumber and expectedNumber are NOT close to each others with a margin of epsilon."		
```
```suggestion
	"Tell whether the actualNumber and expectedNumber are NOT close to each others with a margin of epsilon."
```

juliendelplanque commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -404,15 +377,14 @@ TestAsserter >> denyCollection: actualCollection equals: expectedCollection [
 
 { #category : #asserting }
-TestAsserter >> denyCollection: actualCollection includesAll: subcollection [
+TestAsserter >> denyCollection: actualCollection includesAll: subcollection
```
```suggestion
TestAsserter >> denyCollection: actualCollection includesAll: subcollection [
```

juliendelplanque reviewed (commented):
Continue to fix stupid merge problems...

juliendelplanque committed:
Update src/SUnit-Core/TestAsserter.class.st

juliendelplanque committed:
Update src/SUnit-Core/TestAsserter.class.st

juliendelplanque committed:
Update src/SUnit-Core/TestAsserter.class.st

juliendelplanque committed:
Apply suggestions from code review

juliendelplanque commented on src/SUnit-Core/TestAsserter.class.st:
```diff
@@ -412,7 +385,6 @@ TestAsserter >> denyCollection: actualCollection includesAll: subcollection [
 	^ self
 		assert: (actualCollection includesAll: subcollection) not
 		description: [ actualCollection asString , ' does not include all in ' , subcollection asString ]
```
```suggestion
		description: [ actualCollection asString , ' does not include all in ' , subcollection asString ]
]
```

juliendelplanque committed:
Update src/SUnit-Core/TestAsserter.class.st

jecisc commented:
There are 18 failures. I'm not sure it is related to this PR so I relaunched the CI.

jecisc commented:
CI crashed. I relaunched it.

juliendelplanque committed:
Merge branch 'Pharo8.0' into 3662-Reify-type-of-test-results

juliendelplanque committed:
Delete .DS_Store
